### ISSUE TITLE
lag with perspective and larger cube sizes

### ISSUE THREAD

**Issue by quantenzitrone** [leonmavr/retrocube#9]
Since I updated retrocube last the command
```cmd
cube -wi 200 -de 200 -he 200 -sx 0.99 -sy 0.90 -up
```
doesn't work properly anymore
on commit `76a13dc378c1dd611819e504d9ba6d00f9649754`:

https://github.com/leonmavr/retrocube/assets/74491719/ac12edc5-e01d-4dc3-bed2-8832372d38bf

on current `HEAD` (`95bf2f54ff98813921d34723afb3f467dc912ad3`):

https://github.com/leonmavr/retrocube/assets/74491719/d472f986-8d85-49bd-8d24-008d7727ddec

The cube offset in the videos is just because I recorded my terminal window on the left with obs and somehow obs still recorded in 1920x1080 

i also tried to record with asciinema, but it didn't work well

**Comment by quantenzitrone** [comment]
It seems like `-up` in combination with a large cube like `-wi 200 -de 200 -he 200` results in the lag.
The drift seems to happen even with no arguments.

**Comment by leonmavr** [comment]
Hey @Quantenzitrone,

It's true that the arguments `-wi 200 -de 200 -he 200` make it slower. That's because `-up` activates the perspective transform and combined with large dimensions this makes it map more pixels from 3D to 2D.

For the drift, it happens because I now bounce the cube around the screen to show off that it can move as you can see below:
https://github.com/leonmavr/retrocube/blob/master/main.c#L44
So I should update the README to show it.

I now have some more demos in the demo folder:
https://github.com/leonmavr/retrocube/tree/master/demos

I haven't worked on this project in summer because I was busy with real life but I'll get back to it in about a week. I'll fix demo 6 and then I'm planning to use path tracing instead of ray tracing to reduce the lag (hopefully!).

**Comment by quantenzitrone** [comment]
So is there currently a way to turn off the bouncing?

**Comment by leonmavr** [comment]
No, it's hardcoded at the latest main. You think I should add a toggle or something?

**Comment by quantenzitrone** [comment]
yes

**Comment by leonmavr** [comment]
The latest master now includes a toggle to enable drift (`-b N`, where N is non-zero number of frames) and some options to define the motion. Drift is off my default. I also updated the README.

**Comment by quantenzitrone** [comment]
beautiful, thank you
the new implementation of perspective still lags very much for larger cube sizes, so i'll keep this open

**Comment by leonmavr** [comment]
Yeah, we can keep it open till I implement the new algorithm

**Comment by leonmavr** [comment]
It's been a while and I forgot about this issue of lag with perspective and large sizes. An easy way to reduce it it to use lookup tables for trigonometry instead of library calls. I switched to computing the trigonometric functions using lookup tables instead of library calls (0e777d7, b9435e0) since perspective makes heavy use of them. It didn't manage to improve the performance a lot but it made an improvement nontheless.

To compare the performance gain I disabled the `nanosleep` call and ran:
```
time ./cube -sx 1 -sy 1 -sz 1 --use-perspective --width 300 --height 300 --depth 100 --max-iterations 200
```

master at d9cfe32 (after LUTs):
```
real	1m22.723s
user	1m22.523s
sys	0m0.044s
```
189bfc1 (before LUts):
```
real	1m43.856s
user	1m43.738s
sys	0m0.024s
```
To improve the performance of the perspective even more I'll need to make some major refactoring and I'm not planning to do it yet so let's close the issue.
